api/scholarqa/scholar_qa.py

```diff
@@ -206,6 +212,164 @@ def step_gen_iterative_summary(self, query: str, per_paper_summaries: Dict[str,
             logger.info(f"Step 3 done, cost: {return_val.tot_cost}, time: {time() - start:.2f}")
         return return_val
 
+    @staticmethod
+    def passage_to_quotes_metadata(retrieval_df: pd.DataFrame, per_paper_summaries: Dict[str, str],
+                                   plan_json: Dict[str, List[int]]) -> Dict[str, List[Dict[str, Any]]]:
+        """
+        Map the quotes extracted in step 1 `per_paper_summaries` to their actual full length versions
+        in retrieval_df and make a map of quote ref string --> [inline citations].
+
+        i) First parse the plan json to get the required paper indices to be sent for the final generation in step 3.
+        ii) For those papers, find the quotes from per_paper_summaries and passages from retrieval_df.
+        iii) For each quote, first check if the paper metadata from retrieval_df consists of passages, if not, then the quotes are taken from
+        abstracts of keyword results.
+        iii) Else, try to match the paper quotes as substrings to the corresponding paper passages to get the inline citations along with
+        offsets.
+        iv) First try matching the raw string (high precision), if unsuccessful, try matching only the alphabet characters (high recall).
+        v) Once a match is found, use its offsets in the passage and iterate over the sentence offsets/inline citations to find any that occur within the snippet.
+        vi) In case of a raw string match, replace any citation mentions (if possible), with the paper id of the corresponding inline citation to be linked later.
+        """
+        # get all the ref strings for the clutering plan generated in step 2
+        ref_str_list = [k for k in per_paper_summaries]
+        # paper identifiers for the selected quotes in the plan obtained from their corresponding index
+        req_ref_strs = {ref_str_list[item] for sublist in plan_json.values() for item in sublist if
+                        item < len(ref_str_list)}
```

Comment:
This method is called after step 2. We first scope the quotes down to only the ones included in the clustering plan.